torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+    """Holds the rendezvous state synced with other nodes."""
+
+    state: _RendezvousState
+
+    def __init__(self) -> None:
+        self.state = _RendezvousState()
+
+    @abstractmethod
+    def sync(self) -> None:
+        """Gets or sets the latest rendezvous state."""
+
+    @abstractmethod
+    def mark_dirty(self) -> None:
+        """Marks the local rendezvous state as dirty."""
+
+
+class _RendezvousStateHolderImpl(_RendezvousStateHolder):
+    """Represents the default implementation of ``RendezvousStateHolder``.
+
+    Args:
+        backend:
+            The rendezvous backend to use.
+        settings:
+            The rendezvous settings.
+        cache_duration:
+            The amount of time, in seconds, to cache the last rendezvous state
+            before requesting it from the backend again.
+    """
+
+    backend: RendezvousBackend
+    settings: RendezvousSettings
+    cache_duration: int
+
+    _token: Token
+    _dirty: bool
+    _last_sync_time: float
+
+    def __init__(
+        self, backend: RendezvousBackend, settings: RendezvousSettings, cache_duration: int = 1
+    ) -> None:
+        super().__init__()
+
+        self.backend = backend
+        self.settings = settings
+        self.cache_duration = cache_duration
+
+        self._token = None
+        self._dirty = False
+        self._last_sync_time = -1
+
+    def sync(self) -> None:
+        """See base class."""
+        if self._dirty:
+            state_bits = pickle.dumps(self.state)
+
+            response = self.backend.set_state(state_bits, self._token)
+        else:
+            if self.cache_duration > 0:
+                # Avoid overloading the backend if we are asked to retrieve the
+                # state repeatedly. Try to serve the cached state.
+                if self._last_sync_time >= max(time.monotonic() - self.cache_duration, 0):
+                    return
+
+            response = self.backend.get_state()
+
+        if response:
+            state_bits, token = response
+
+            try:
+                self.state = pickle.loads(state_bits)
+            except pickle.PickleError as exc:
+                raise RendezvousStateError(
+                    "The rendezvous state is corrupt. See inner exception for details."
+                ) from exc
+        else:
+            token = None
+
+            self.state = _RendezvousState()
+
+        self._token = token
+        self._dirty = False
+
+        self._last_sync_time = time.monotonic()
+
+        self._sanitize()
+
+    def _sanitize(self) -> None:
+        expire_time = datetime.utcnow() - (
+            self.settings.keep_alive_interval * self.settings.keep_alive_max_attempt
+        )
+
+        # Filter out the dead nodes.
+        dead_nodes = [
+            node
+            for node, last_keep_alive in self.state.last_keep_alives.items()
+            if last_keep_alive < expire_time
+        ]
+
+        for dead_node in dead_nodes:
```

Comment:
If we really need to remove dead_nodes, we probably would like to at least have some logs around what nodes got removed and from what node.

Reply:
Makes sense. I will add some logging here.